### PR TITLE
Unsupported browser (feature)

### PR DESCRIPTION
**Pull request by BeatC**

### CODE REVIEW

File: react/features/app/actions.js [hunk 1]
```diff
@@ -6,11 +6,20 @@ import { APP_WILL_MOUNT, APP_WILL_UNMOUNT } from './actionTypes';
...
 } from './functions';
 import './reducer';
 
 /**
+ * Variable saving current route of the app. Later it will be extracted as
+ * a property of the class with Router specific logic.
+ *
+ * @type {Object}
+ */
+let currentRoute = {};
```
Comment (lyubomir) [inline react/features/app/actions.js]: 1. No global state please! Either React Component state or Redux store.

2. AbstractApp's `state` already tracks the current route. It has a `route` property.

File: react/features/app/actions.js [hunk 2]
```diff
@@ -147,15 +147,15 @@ function _navigate(state) {
...
         dispatch(setRoom(newRoom));
 
         const state = getState();
-        const room = state['features/base/conference'].room;
+        const { app } = state['features/app'];
+        const newRoute = _getRouteToRender(state);
 
-        if (room !== oldRoom) {
-            _navigate(state);
+        if (!areRoutesEqual(newRoute, currentRoute)) {
```
Comment (lyubomir) [inline react/features/app/actions.js]: `RouteRegistry` clones routes before it exposes them from its internal state so `areRoutesEqual` would probably better fit as a (static?) method of `RouteRegistry`.

File: react/features/app/actions.js [hunk 3]
```diff
@@ -147,15 +147,15 @@ function _navigate(state) {
  */
 function _setRoomAndNavigate(newRoom) {
     return (dispatch, getState) => {
-        const oldRoom = getState()['features/base/conference'].room;
-
         dispatch(setRoom(newRoom));
 
         const state = getState();
-        const room = state['features/base/conference'].room;
+        const { app } = state['features/app'];
```
Comment (lyubomir) [inline react/features/app/actions.js]: The source code from this line to the end of the method would probably better fit in `App`'s `_navigate` method.

File: react/features/unsupported-browser/components/Landing.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    android: 'https://play.google.com/store/apps/details?id=org.jitsi.meet',
+    ios: 'https://itunes.apple.com/us/app/jitsi-meet/id1165103905'
+};
+
+/**
+ * React component representing mobile landing page.
+ *
+ * @class Landing
+ */
+class Landing extends Component {
```
Comment (lyubomir) [inline react/features/unsupported-browser/components/Landing.js]: Please replace the word landing in all its occurrences (class, Redux state, etc.) with something more appropriate given that the feature is named `unsupported-browser`.

File: react/features/app/actions.js [hunk 5]
```diff
@@ -59,7 +69,15 @@ export function appNavigate(urlOrRoom) {
             loadConfig(`https://${domain}`)
                 .then(
                     config => configLoaded(/* err */ undefined, config),
-                    err => configLoaded(err, /* config */ undefined));
+                    err => configLoaded(err, /* config */ undefined))
+                .then(() => {
+                    const link = typeof room === 'undefined'
+                    && typeof domain === 'undefined'
+                        ? urlOrRoom
+                        : room;
```
Comment (lyubomir) [inline react/features/app/actions.js]: The formatting of the above is not in accord with the existing source code.

File: react/features/app/functions.web.js [hunk 6]
```diff
@@ -9,7 +16,53 @@ import JitsiMeetLogStorage from '../../../modules/util/JitsiMeetLogStorage';
...
+    const { landingIsShown } = state['features/unsupported-browser'];
+    const { room } = state['features/base/conference'];
+    const component = isRoomValid(room) ? Conference : WelcomePage;
+
+    // We're using spread operator here to create copy of the route registered
+    // in registry. If we overwrite some of its properties (like 'component')
+    // they will stay unchanged in the registry.
+    const route = { ...RouteRegistry.getRouteByComponent(component) };
+
+    if ((OS === 'android' || OS === 'ios') && !landingIsShown) {
```
Comment (lyubomir) [inline react/features/app/functions.web.js]: The check should be whether the execution environment supports the app. The biggest piece of it is probably whether WebRTC is supported.

Only then does come the refinement that we choose to promote the mobile app in Google Chrome (if it supports WebRTC) on Android.

File: react/features/unsupported-browser/components/Landing.js (hunk 4 above)
Comment (BeatC) [inline react/features/unsupported-browser/components/Landing.js]: Renamed `Landing` component and logic related to that to `MobileBrowserPage`

File: react/features/app/actions.js (hunk 1 above)
Comment (BeatC) [inline react/features/app/actions.js]: Removed global state.

**Comment by BeatC** [comment]
jenkins test this please

**Pull request merged**